ENH: stats.moment: add array API support

mdhaber opened the pull request:
#### Reference issue
Towards gh-18867

#### What does this implement/fix?
Adds array API support to `stats.moment`.

mdhaber committed:
ENH: stats.moment: add array API support

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -1017,26 +1018,32 @@ def moment(a, order=1, axis=0, nan_policy='propagate', *, center=None):
-    if np.abs(moment - np.round(moment)) > 0:
+    xp = array_namespace(a)
+
+    if xp.abs(moment - xp.round(moment)) > 0:
```
```suggestion
    if np.abs(moment - np.round(moment)) > 0:
```

mdhaber committed:
MAINT: stats._moment: fix bugs

[skip ci]

lucascolley commented:
> Is there something I need to do to get this to run with alternative backends in CI?

Add a new pytest call in the array API CI workflow. `-s stats` might be overkill for now, maybe you can make it a little more fine grained with `-t`.

lucascolley reviewed (commented):
just a few comments for now

mdhaber commented on scipy/_lib/_util.py:
```diff
@@ -708,23 +708,27 @@ def _nan_allsame(a, axis, keepdims=False):
 
-    if np.issubdtype(a.dtype, np.inexact):
-        # The summation method avoids creating a (potentially huge) array.
+    inexact = ('float' in str(a.dtype) or 'complex' in str(a.dtype))
```
Cool. Didn't see that.

lucascolley commented on scipy/_lib/tests/test__util.py:
```diff
@@ -331,6 +331,33 @@ def test_contains_nan_2d(self):
+        rng = np.random.default_rng(932347235892482)
+        x0 = rng.random(size=(2, 3, 4))
+        x = xp.asarray(x0)
+        x_nan = xp.asarray(x0.copy())
```
Oops, I meant `copy(x, xp=xp)`. 1 less `np`->`xp` means 1 less potential device transfer

mdhaber replied:
Did you mean save a copy? I don't think we can save a device transfer here, since `x0` is guaranteed to be generated by NumPy on the CPU, but we want it on whatever device `xp` is using. 

It looks like this could save a `copy`, though: if the array already needs to be copied from the CPU to another device, we don't need to copy the NumPy array. In that case, this `copy` function is what I was looking for, because what I really wanted was `xp.asarray(x0, copy=True)` to ensure that exactly one copy is performed.

mdhaber replied:
I see. You wrote `copy(x0, xp=xp)` above.

mdhaber committed:
Merge remote-tracking branch 'upstream/main' into moment_array_api

mdhaber committed:
MAINT: stats._moment: improve array API compatibility

mdhaber committed:
TST: stats._moment: run tests on CI, skip failing tests

mdhaber committed:
MAINT:stats.moment: rename variable 'moment' to 'order' throughout

mdhaber committed:
ENH: stats.moment: add array API support

mdhaber committed:
MAINT: stats._axis_nan_policy: skip when using non-NumPy backend

lucascolley commented on scipy/_lib/_util.py:
```diff
@@ -709,22 +709,29 @@ def _nan_allsame(a, axis, keepdims=False):
 def _contains_nan(a, nan_policy='propagate', use_summation=True,
                   policies=None):
-    if not isinstance(a, np.ndarray):
+    # _contains_nan expects `xp` to be an array-api compatible namespace
+    # (e.g. from scipy._lib.array_api_compat).
+    xp = array_namespace(a)
```
Can you follow the pattern of allowing the option to pass `xp` around private functions as a kwarg to avoid duplicate `array_namespace` calls:

```python
def _contains_nan(a, nan_policy='propagate', use_summation=True,
                  policies=None, *, xp=None):
	if xp is None:
		xp = array_namespace(a)
```

mdhaber replied:
Probably. This was done in pieces with the lowest level functions first, so the higher-level functions weren't passing `xp` yet.

mdhaber replied:
Done.

lucascolley commented on scipy/stats/_axis_nan_policy.py:
```diff
@@ -4,12 +4,15 @@
 from functools import wraps
 from scipy._lib._docscrape import FunctionDoc, Parameter
 from scipy._lib._util import _contains_nan, AxisError, _get_nan
+from scipy._lib._array_api import array_namespace, is_numpy
 import inspect
 
+_SCIPY_ARRAY_API = os.environ.get("SCIPY_ARRAY_API", False)
```
Do you need the env variable here? If so, no need to pull from `os.environ` twice.

```suggestion
from scipy._lib._array_api import array_namespace, is_numpy, SCIPY_ARRAY_API
import inspect
```

mdhaber replied:
OK, I can pull from `scipy._lib._array_api` instead.

I was using `pytest.mark.skipif(SCIPY_ARRAY_API)` to skip tests that pass masked arrays, object arrays, etc.

mdhaber replied:
Done.

lucascolley commented on scipy/stats/_stats_py.py:
```diff
@@ -100,16 +101,16 @@
 
 
-def _chk_asarray(a, axis):
+def _chk_asarray(a, axis, xp=np):
```
I think I would prefer `_chk_asarray(a, axis, *, xp=None)` with
```python
if xp is None:
	xp = array_namespace(a)
```
but let me know if you disagree.

mdhaber replied:
My plan was to always pass `xp` when updating functions to use the array API. For any function that doesn't pass `xp` when it calls `_chk_asarray`, `np` is correct, so it seems unnecessary to call `array_namespace` to determine that.

lucascolley replied:
Alright. I have heard others object to having `np` as a default value though - a pattern used in cluster is `xp=None`, with `xp = np if xp is None else xp`,

mdhaber replied:
I went ahead and changed it.

lucascolley commented on scipy/stats/tests/test_axis_nan_policy.py:
```diff
@@ -16,6 +17,8 @@
 from scipy.stats._axis_nan_policy import _masked_arrays_2_sentinel_arrays
 from scipy._lib._util import AxisError
 
+_SCIPY_ARRAY_API = os.environ.get("SCIPY_ARRAY_API", False)
```
same comment about importing from `_array_api`

mdhaber replied:
Done.

mdhaber commented:
@lucascolley Thanks for the quick reviews - I'm glad you're doing that! Just know that you will find a lot of things if you get to it before I am able to review my own code on GitHub. Seeing just the diff highlights these things, which I might not be so careful when drafting the code.

Re: importing `SCIPY_ARRAY_API` - the reason I'm using it is that I think I have a legitimate need for skipping tests whenever `SCIPY_ARRAY_API == 1`. The tests involve masked and/or object arrays, but `array_namespace` raises errors when it gets these. 

I think this will be common enough that it does make sense to have a custom mark that really skips the test whenever `SCIPY_ARRAY_API == 1`, like `skip_if_array_api` used to. Otherwise, as you see in the CI logs, I _have_ to pass a `reason` when using `pytest.mark.skipif`, and that is going to get really repetitive when the reason is always the same. Does it make sense to have a real `skip_if_array_mark` after all?

mdhaber committed:
TST: stats.moment: begin to convert existing tests to array API

mdhaber commented on scipy/_lib/_util.py:
```diff
@@ -708,23 +708,31 @@ def _nan_allsame(a, axis, keepdims=False):
-                  policies=None):
-    if not isinstance(a, np.ndarray):
+                  policies=None, *, xp=None):
+    # _contains_nan expects `xp` to be an array-api compatible namespace
+    # (e.g. from scipy._lib.array_api_compat).
```
```suggestion
```

mdhaber commented on scipy/_lib/tests/test__util.py:
```diff
@@ -1,3 +1,4 @@
+import os
```
Ugh missed this one.

mdhaber committed:
TST: scipy._lib._contains_nan: fixup

mdhaber committed:
MAINT: stats._axis_nan_policy: fixup decorator early exit with non-numpy arrays

mdhaber commented on scipy/_lib/_util.py:
```diff
@@ -708,23 +708,29 @@ def _nan_allsame(a, axis, keepdims=False):
-        # The summation method avoids creating a (potentially huge) array.
+    inexact = (xp.isdtype(a.dtype, "real floating")
+               or xp.isdtype(a.dtype, "complex floating"))
+    if inexact or (is_numpy(xp) and np.issubdtype(a.dtype, np.inexact)):
```
```suggestion
    if inexact:
```

lucascolley commented:
> I think this will be common enough that it does make sense to have a custom mark like the original skip_if_array_api. Otherwise, as you see in the CI logs, I have to pass a reason when using pytest.mark.skipif, and that is going to get really repetitive when the reason is always the same. Does it make sense to have a real skip_if_array_mark after all?

@rgommers you originally wanted this marker removed, but I think it is the best way to deal with tests that involved masked/object/`numpy.matrix` arrays. Thoughts?

rgommers commented:
> @rgommers you originally wanted this marker removed, but I think it is the best way to deal with tests that involve masked/object/`numpy.matrix` arrays. Thoughts?

I'm thinking about what will happen if and when we make the behavior proposed in gh-18286 the default. If something is simply skipped now, it may silently stop being tested or be blocking for changing the default behavior. I see that I already anticipated the issue with masked array support in `stats.mstats` in gh-18286, see the entry for `numpy.ma.MaskedArray` in the first table there. That suggests:
- `stats.mstats` functions should return masked arrays when they receive masked arrays as inputs
- `stats` functions should reject masked arrays
- `object` arrays are never accepted, start raising `TypeError`
- `numpy.matrix` instances are never accepted, start raising `TypeError`

For the current behavior, yes it seems like there should be a way to skip tests that don't comply with the above. Ideally in a way that makes it easy to move over to the new defaults without anything going wrong silently.

> I have to pass a reason when using pytest.mark.skipif, and that is going to get really repetitive when the reason is always the same

`skip_if_array_api` isn't really a reason, so I'm slightly worried about using that. The array API code path already raises a hard error, so how are we going to add deprecations for behavior that you'd want to skip with `skip_if_array_api`? It seems to me like we'd need separate code paths inside the functions for "doesn't work with array-api defaults", and that it'd be useful to have more specific skip markers like `accepts_object_arrays`, `accepts_npmatrix`. Does that make sense here?

mdhaber commented:
> The array API code path already raises a hard error, so how are we going to add deprecations for behavior that you'd want to skip with `skip_if_array_api`? 

I think I'm not seeing the problem. Suppose there were a decorator `skip_perfect_name_here` that skips tests when `SCIPY_ARRAY_API=1`. This would be applied to tests that pass array API incompatible input to the function being tested. That is, it would be applied to tests that would otherwise fail because `array_namespace` raises an error (whether that be due to receiving an object array, a NumPy matrix, a masked array, or something else). 

Immediately, when the function is called with array API incompatible input:
- the function runs normally when `SCIPY_ARRAY_API` is not specified[^1], so the test continues to do its job.
- the function raises an error when `SCIPY_ARRAY_API=1`, but the test is skipped.

Sometime before the behavior proposed in gh-18286 becomes the default, we'll functions that currently accept array API invalid input need begin emitting deprecation warnings when they receive array API invalid input. This can be a separate effort from adding the array API functionality, and it can be done after adding array API functionality. To be honest, `array_namespace` could begin to emit the deprecation warning when it receives array API incompatible input and `SCIPY_ARRAY_API` is not set.[^2]

At this point, when the function is called with array API incompatible input:
- the function emits a deprecation warning when `SCIPY_ARRAY_API` is not specified, causing tests marked with `skip_perfect_name_here` to fail. We can either manually update the tests at that time to look for (and not fail due to) the deprecation warning, or maybe `skip_perfect_name_here` can be updated to do that automatically.
- the function raises an error when `SCIPY_ARRAY_API=1`, so it will still be skipped.

When we decide to pull the trigger and make the behavior proposed in gh-18286 (or something similar) the default and only behavior, we remove the decorator. Tests to which the decorator has been applied will begin to fail, but they are no longer relevant, so we can remove them. If there is something in these tests that is still relevant, we move those parts into separate tests.

I don't mean for this to be a real proposal; I just mean to be explicit enough about a potential path forward so we can identify problems with having a `skip_perfect_name_here` decorator. 

[^1]: IIRC `SCIPY_ARRAY_API=0` doesn't work right now
[^2]: One potential pitfal would be that it would also emit a deprecation warning for functions that *do not* currently work with array API invalid input, suggesting that the *do* but won't in the future. I don't think this is a problem provided that the warning message is worded appropriately.

lucascolley commented:
I think what Matt has written makes sense.

> To be honest, array_namespace could begin to emit the deprecation warning when it receives array API incompatible input and SCIPY_ARRAY_API is not set.

+1, I think we can word the warnings appropriately to avoid the concern in your footnote.

mdhaber committed:
TST: stats.moment: make tests array_api_compatible

mdhaber committed:
TST: stats: adjust some skipped tests

[skip cirrus] [skip circle]

mdhaber commented on scipy/_lib/tests/test__util.py:
```diff
@@ -302,7 +305,7 @@ def test_policy(self):
             _contains_nan(data, nan_policy="nan")
 
-    def test_contains_nan_1d(self):
+    def test_contains_nan(self):
```
Just moved tests with non-numeric input into a separate function so they could be skipped.

mdhaber commented on scipy/stats/_axis_nan_policy.py:
```diff
@@ -8,9 +8,9 @@
 from scipy._lib._util import _contains_nan, AxisError, _get_nan
+from scipy._lib._array_api import array_namespace, is_numpy
 import inspect
 
```
```suggestion


```
Line below removed inadvertently.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -889,11 +894,11 @@ def tsem(a, limits=None, inclusive=(True, True), axis=0, ddof=1):
 def _moment_outputs(kwds):
-    moment = np.atleast_1d(kwds.get('order', 1))
-    if moment.size == 0:
+    order = np.atleast_1d(kwds.get('order', 1))
```
I probably shouldn't have complicated this PR by renaming the old variable `moment` to `order`... but I did.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -998,51 +1003,68 @@ def moment(a, order=1, axis=0, nan_policy='propagate', *, center=None):
 
 # Moment with optional pre-computed mean, equal to a.mean(axis, keepdims=True)
-def _moment(a, moment, axis, *, mean=None):
-    if np.abs(moment - np.round(moment)) > 0:
-        raise ValueError("All moment parameters must be integers")
```
This is checked in the public function now.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -998,51 +1003,68 @@ def moment(a, order=1, axis=0, nan_policy='propagate', *, center=None):
-    # moment of empty array is the same regardless of order
-    if a.size == 0:
-        return np.mean(a, axis=axis)
+    if xp.isdtype(a.dtype, 'integral'):
+        a = xp.asarray(a, dtype=xp.float64)
```
When `skew`, `kurtosis`, and any other functions that rely on `_moment` are updated, this can be removed, since it's already in the public function.

rgommers commented:
> I don't necessarily mean for this to be a real proposal, but hopefully it's explicit enough about a potential path forward so we can identify problems with having a `skip_perfect_name_here` decorator or mark.

Thanks Matt for thinking it through. I agree, that should work as a general strategy.

The proposed name should work, but some hints as to why would be helpful, at least in the first usage in a set of functions. If we don't want specific decorators, then something like:
```python
@skip_if_array_api   # uses masked arrays
```
it's so much easier than reading the code to deduce what the reason for a skip is

mdhaber commented:
Thanks @rgommers. Does the reason need to be more specific than that the tested function is passed an array API invalid input, or do all the different types of invalid inputs used need to be specified? 

If the former,  would `skip_array_api_invalid_arg` without comment work?  

And just to double check, should this be a decorator that applies a mark so that there is potential for the decorator to do more in the future (e.g. look for deprecation warnings, if that's possible)?

rgommers commented:
> If the former, would `skip_array_api_invalid_arg` without comment work, if this plan is documented in the contributor guide?

That sounds good to me. In most cases it'll be clear enough with the test code right below. In case it's non-obvious (e.g. function generates object array internally), an extra comment as to why may be helpful.



> And just to double check, should this be a decorator that applies a mark so that there is potential for the decorator to do more in the future (e.g. look for deprecation warnings, if that's possible)?

I think so - that's the most efficient way of doing it anyway, right?

mdhaber commented:
Definitely. The only question is whether we would want to be more fine grained when looking for deprecation warnings. Some tests will have multiple calls to the function, and the decorated pass would not distinguish between the deprecation warning being emitted between different calls. But I think it would be safe enough to assume that all of them emit the warning if one of them does, especially if the warning is actually emitted by something central like `array_namespace`.

rgommers commented:
Agreed, that seems safe enough to assume.

mdhaber committed:
Merge remote-tracking branch 'upstream/main' into moment_array_api

mdhaber committed:
TST: define skip_array_api_invalid_arg decorator

mdhaber committed:
DOC: document skip_array_api_invalid_arg decorator

mdhaber committed:
MAINT: stats.moment: minor fixups

mdhaber commented on scipy/_lib/_util.py:
```diff
@@ -708,23 +708,29 @@ def _nan_allsame(a, axis, keepdims=False):
+    inexact = (xp.isdtype(a.dtype, "real floating")
+               or xp.isdtype(a.dtype, "complex floating"))
+    if inexact:
+        # The summation method avoids creating another (potentially huge) array
```
However, it can introduce NaNs where non existed before (see gh-20386). I propose getting rid of it and always use the other method: `contains_nan = xp.isnan(xp.sum(a))`.

mdhaber commented:
@lucascolley besides incorporating a fix for gh-20386, is there anything else you'd like to see here?

lucascolley commented on scipy/conftest.py:
```diff
@@ -151,6 +151,10 @@ def check_fpu_mode(request):
 
 array_api_compatible = pytest.mark.parametrize("xp", xp_available_backends.values())
 
+skip_array_api_invalid_arg = pytest.mark.skipif(SCIPY_ARRAY_API,
```
can we rename this to `skip_xp_invalid_arg`, it's quite long and that is just as clear IMO.

mdhaber replied:
Fine with me. @rgommers?

mdhaber replied:
Renamed

lucascolley commented on scipy/conftest.py:
```diff
@@ -151,6 +151,10 @@ def check_fpu_mode(request):
 
+skip_array_api_invalid_arg = pytest.mark.skipif(SCIPY_ARRAY_API,
+    reason = ('Test involves masked arrays, object arrays, or other '
+              'types that are not valid input when SCIPY_ARRAY_API=1'))
```
not _quite_ true, given that this is the definition of `SCIPY_ARRAY_API`:

```
SCIPY_ARRAY_API: str | bool = os.environ.get("SCIPY_ARRAY_API", False)
```

I would just say "when `SCIPY_ARRAY_API` is set".

mdhaber replied:
Ok, but is it intended for `SCIPY_ARRAY_API = 0` to have the same behavior as `1`? I've noted elsewhere that `SCIPY_ARRAY_API = 0` results in errors, but I've assumed this is a mistake.

mdhaber replied:
In that case, let's not change this documentation. Setting it to `0` doesn't and should not do the same as setting it to `1`, and I don't know that we'd want to advertise that it can be any truthy value if it's supposed to be a boolean.
I can file an issue about the error when it is set to `0`.

mdhaber replied:
I misunderstood originally - if the distinction was that "True" is also a valid input, I don't think we would need to change this. However, `SCIPY_ARRAY_API` can be JSON, which is different than `True`, but presumably it also causes xp-incompatible input to fail. So I changed the wording to "types that are not valid input when `SCIPY_ARRAY_API` is _used_", which leaves a little room for interpretation compared to "set".

lucascolley replied:
Yes, that sounds better than what I had in mind, thanks.

lucascolley commented on scipy/stats/_stats_py.py:
```diff
@@ -999,51 +1003,68 @@ def moment(a, order=1, axis=0, nan_policy='propagate', *, center=None):
+    if xp_size(order) == 0:
+        # This is tested by `_moment_outputs`, which is run by the `_axis_nan_policy`
+        # decorator. Currently, the `_axis_nan_policy` decorator is skipped when `a`
+        # is a non-NumPy array, so we need to check again. When the decoratpor is
```
```suggestion
        # is a non-NumPy array, so we need to check again. When the decorator is
```

mdhaber replied:
Lumped this in with other changes locally rather than committing here.

lucascolley commented on scipy/stats/_stats_py.py:
```diff
@@ -999,51 +1003,68 @@ def moment(a, order=1, axis=0, nan_policy='propagate', *, center=None):
+        return _moment(a, order, axis, mean=center)
 
 
 # Moment with optional pre-computed mean, equal to a.mean(axis, keepdims=True)
```
```suggestion
# Moment with optional pre-computed mean, equal to `xp.mean(a, axis=axis, keepdims=True)`
```

mdhaber replied:
Same as above but this comment didn't make sense to me, so I rewrote it as a minimal docstring.

lucascolley commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -235,10 +238,22 @@ The following pytest markers are available:
   ``@pytest.mark.usefixtures("skip_xp_backends")`` must be used alongside this
   marker for the skipping to apply.
+* ``skip_array_api_invalid_arg`` is used to skip tests that use arguments which
+  are invalid when ``SCIPY_ARRAY_API=1``. For instance, some tests of
```
same comment as above/below about "`SCIPY_ARRAY_API` is set".

lucascolley commented on doc/source/dev/api-dev/array_api.rst:
```diff
@@ -235,10 +238,22 @@ The following pytest markers are available:
+  decorator will check that the deprecation warning is emitted without it
+  causing the test to fail. When ``SCIPY_ARRAY_API=1`` behavior becomes the
+  default and only behavior, these tests (and the decorator itself) will be
+  removed.
```
nice description, thanks!

lucascolley commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -190,7 +191,8 @@ def test_empty_input(self):
     def test_not_enough_values(self):
         assert_raises(ValueError, stats.shapiro, [1, 2])
-        assert_raises(ValueError, stats.shapiro, np.array([[], [2]], dtype=object))
+        error_type = TypeError if SCIPY_ARRAY_API else ValueError
+        assert_raises(error_type, stats.shapiro, np.array([[], [2]], dtype=object))
```
👍 I think this is the right pattern to use

lucascolley commented on scipy/_lib/tests/test__util.py:
```diff
@@ -10,9 +10,9 @@
+from scipy.conftest import array_api_compatible, skip_array_api_invalid_arg
 
-from scipy._lib._array_api import xp_assert_equal
+from scipy._lib._array_api import xp_assert_equal, is_numpy, copy as xp_copy
```
I'm not opposed to changing the names of the helpers themselves to `xp_copy` and `xp_size` if that would be easier than having these aliases everywhere. `size` is just exposed from `array_api_compat` (no wrapper), and `copy` was written by Andrew.

lucascolley commented on scipy/_lib/tests/test__util.py:
```diff
@@ -331,6 +332,33 @@ def test_contains_nan_2d(self):
 
+    @array_api_compatible
+    @pytest.mark.parametrize("nan_policy", ['propagate', 'omit', 'raise'])
+    def test_array_api(self, xp, nan_policy):
```
is this similar to the other PR in that this test is here in place of pending improvements to the other tests, or is it here to stay?

mdhaber replied:
Similar to `pearsonr`.

lucascolley commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3230,112 +3234,163 @@ def _assert_equal(self, actual, expect, *, shape=None, dtype=None):
         res = stats.moment(x, m, center=c)
-        ref = [stats.moment(x, i, center=c) for i in m]
-        assert_equal(res, ref)
+        xp = array_namespace(x)
+        ref = xp.concat([stats.moment(x, i, center=c)[xp.newaxis, ...] for i in m])
```
Just to help mark that verbosity can be reduced once our minimum NumPy version is 2.0.

```suggestion
        xp_test = array_namespace(x) # np<2.0 does not have `xp.concat`
        ref = xp_test.concat([stats.moment(x, i, center=c)[xp.newaxis, ...] for i in m])
```

lucascolley commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3230,112 +3234,163 @@ def _assert_equal(self, actual, expect, *, shape=None, dtype=None):
+    def test_moment_accuracy(self):
+        # 'moment' must have a small enough error compared to the slower
+        #  but very accurate numpy.power() implementation.
+        tc_no_mean = self.testcase_moment_accuracy - \
+                     np.mean(self.testcase_moment_accuracy)
```
nit: I would prefer this but feel free to ignore
```suggestion
        tc_no_mean = (self.testcase_moment_accuracy
                      - np.mean(self.testcase_moment_accuracy))
```

mdhaber replied:
Not sure how my IDE got away with this. I never use explicit line breaks or put the operator on the previous line.

lucascolley reviewed (changes requested):
almost there!

mdhaber committed:
MAINT: stats.moment: adustments per review

mdhaber commented:
OK, I think the last commit addresses the comments. I will change https://github.com/scipy/scipy/pull/20292#pullrequestreview-1977843015 in a separate PR, as that issue really is orthogonal.

lucascolley reviewed (approved):
looks good from my POV, thanks Matt!

mdhaber commented:
Thanks very much @lucascolley! 
@tupui do you have a moment to skim from a stats perspective?

mdhaber committed:
Merge branch 'main' into moment_array_api

tupui reviewed (approved):
This looks great 👍 The implementation looks correct, same for the tests and the changes for Array API helpful and Lucas checked that part thoroughly 🙌

Feel free to go ahead @lucascolley

lucascolley commented:
thanks @tupui , @mdhaber !

mdhaber commented:
Thanks @lucascolley @rgommers @tupui!
